schema_tables: fix secondary old indexes during their initialization

psarna opened the pull request:
Old secondary index schemas did not have their idx_token column
marked as computed, and there already exists code which updates
them. Unfortunately, the fix itself contains an error and doesn't
fire if computed columns are not yet supported by the whole cluster,
which is a very common situation during upgrades.

Fixes #7515
Tests: unit(dev), manual: adding an index, removing its schema info manually from `system_schema.computed_columns`, then restarting the node and trying to write to the index: it fails before the patch and succeeds after

psarna commented:
/cc @vladzcloudius

avikivity commented on db/schema_tables.cc:
```diff
@@ -2684,6 +2684,20 @@ static index_metadata create_index_from_index_row(const query::result_set_row& r
+            && cdef.kind == column_kind::clustering_key
+            && cdef.id == 0
+            && !cdef.is_computed()
+            && boost::algorithm::ends_with(cf_name, "_index")) {
```
Can we add more conditions, to reduce the probabilty that a user table that happens to be named something_index gets misclassified here?

psarna replied:
The current conditions are:
1. It's a materialized view (this function is only used for views)
2. It has a `idx_token` column ...
3. ... as a first clustering key ...
4. and this view's name ends with `_index`

I think that `2` is the least likely of these, but I wish we named it `__scylla_idx_token` back in the days. This point in code is quite elegant to fix the index, but its one disadvantage is that we can't assume that other tables are loaded yet.

Ideally, we should also check that this table actually corresponds to a secondary index, but that can only be done later. The problem with checking later is that the index is already created, so that involves triggering a schema change announcement locally, and I worry that it's prone to races, because there's a window in which the index schema is incorrect, and also I'm not sure how well such internal schema change requests coexist with waiting for schema agreement, etc.

eliransin replied:
@psarna 

> Ideally, we should also check that this table actually corresponds to a secondary index, but that can only be done later.

I also encountered it as a limitation when working on some other fix, I am starting to think, maybe we should only load materialized views after all "main" tables have been loaded. This can actually solve a lot of races and allow us to have the invariant " If a view exists it's base table exists".

There is still a small nit, I remember seeing this whole piece of code whose purpose is to determine this idx_token name - in the 0.000001% chance that someone had this column as one of it's PK columns :slightly_smiling_face: , in that case I believe there would still be a crash right?
Maybe we need to go for a pattern (_*idx_token)  and a position in the PK (meaning the first clustering key), unless I misunderstood this token purpose, it must be the first clustering key, no?

psarna replied:
Oh, right, purely theoretically, the name can be `idx_token_N`, with `N` being a natural number.

avikivity replied:
I think that the name idx_token is ugly enough. I missed that it was a precondition.

We use the `$` to mark internal identifiers. e.g. `cdc$blah`. So the internal tables could have been named `blah$index`.

eliransin reviewed (commented):
Other than this small nit, LGTM.

psarna commented:
Actually, let me try and test another way - repreparing the schema lazily on error - which means that only after we determine that the token column is not properly computed, we'll first reschedule a schema update and only then continue with the view update path. That way there's 0% chance for falsely marking somebody's regular column as computed. I'll give it a try and send the result here if it looks good.

psarna commented:
Waiiiiiit.... I finally found the code which I added long time ago specifically for this purpose. It's in `schema_tables.cc`:
```cpp
future<> maybe_update_legacy_secondary_index_mv_schema(service::migration_manager& mm, database& db, view_ptr v) {
    // TODO(sarna): Remove once computed columns are guaranteed to be featured in the whole cluster.
    // Legacy format for a secondary index used a hardcoded "token" column, which ensured a proper
    // order for indexed queries. This "token" column is now implemented as a computed column,
    // but for the sake of compatibility we assume that there might be indexes created in the legacy
    // format, where "token" is not marked as computed. Once we're sure that all indexes have their
    // columns marked as computed (because they were either created on a node that supports computed
    // columns or were fixed by this utility function), it's safe to remove this function altogether.
    if (!db.features().cluster_supports_computed_columns()) {
        return make_ready_future<>();
    }

    if (v->clustering_key_size() == 0) {
        return make_ready_future<>();
    }
    const column_definition& first_view_ck = v->clustering_key_columns().front();
    if (first_view_ck.is_computed()) {
        return make_ready_future<>();
    }

    table& base = db.find_column_family(v->view_info()->base_id());
    schema_ptr base_schema = base.schema();
    // If the first clustering key part of a view is a column with name not found in base schema,
    // it implies it might be backing an index created before computed columns were introduced,
    // and as such it must be recreated properly.
    if (!base_schema->columns_by_name().contains(first_view_ck.name())) {
        schema_builder builder{schema_ptr(v)};
        builder.mark_column_computed(first_view_ck.name(), std::make_unique<token_column_computation>());
        return mm.announce_view_update(view_ptr(builder.build()), true);
    }
    return make_ready_future<>();
}
```
Now I just need to figure out how the hell it didn't launch and fix the index...

psarna commented:
Oh boy, I know why...
```cpp
   if (!db.features().cluster_supports_computed_columns()) {
        return make_ready_future<>();
    }
```
we actually want the opposite - during the upgrade, it's very likely that the **cluster** doesn't support computed columns, what we want is to update this column as computed locally. Sigh, the fix is probably a one-liner.

psarna commented:
@avikivity please re-review, this pull request actually makes sense now

avikivity commented on db/schema_tables.cc:
```diff
@@ -3035,10 +3035,6 @@ future<> maybe_update_legacy_secondary_index_mv_schema(service::migration_manage
-    if (!db.features().cluster_supports_computed_columns()) {
-        return make_ready_future<>();
-    }
-
```
Should we protect this with an even newer feature? e.g. cluster_computed_columns_schema_upgraded()?

So that we can skip the code below once it's been upgraded.

I'm not in favor of it myself, such games were just proven to be bad, I'm raising it for your consideration.

psarna replied:
I think there's no reason to - we're talking about a local schema upgrade, which is a glorified way of telling a local node that this column is computed. It will most likely not even get a chance to be propagated, but even if it does, the propagation code is already backward compatible and wouldn't send computed column information to nodes which do not understand it. So the check here was not only incorrect, but totally redundant.

psarna replied:
*just to make sure I'm not talking gibberish, I span a local 3-node cluster and tried to ruin it by removing the computed columns information from some nodes and restarting them, but everything worked properly

psarna replied:
Oh, I might have misunderstood the request. This code runs on startup only, so I think it's better for it to run redundantly than to introduce a new cluster feature.

psarna replied:
Hm, I also think that this code might run before the feature service has a chance to inquire other nodes about their cluster features... in which case this condition was as good as `if (false) { fix_stuff() }`.

avikivity replied:
So, the original code had the condition reversed, as well as evaluated before its inputs were ready? Nice.

psarna replied:
Yeah, I'm particularly proud of that one. I think I'll print, laminate it and hang it on my wall.

avikivity replied:
It's good that something like that can never get past reviewers.

avikivity commented:
Queued.
